usb/device/USBAudio/USBAudio.h

```diff
@@ -0,0 +1,379 @@
+    * @param channel_count_rx channel number (1 or 2) (default: 1)
+    * @param frequency_tx frequency in Hz (default: 8000)
+    * @param channel_count_tx channel number (1 or 2) (default: 1)
+    * @param buffer_ms time audio can be buffered without overflowing in milliseconds
+    * @param vendor_id Your vendor_id
+    * @param product_id Your product_id
+    * @param product_release Your product_release
+    */
+    USBAudio(bool connect=true, uint32_t frequency_rx = 48000, uint8_t channel_count_rx = 1, uint32_t frequency_tx = 8000, uint8_t channel_count_tx = 1, uint32_t buffer_ms=10, uint16_t vendor_id = 0x7bb8, uint16_t product_id = 0x1111, uint16_t product_release = 0x0100);
+
+    /**
+    * Fully featured constructor
+    *
+    * Construct this object with the supplied USBPhy and parameters. The user
+    * this object is responsible for calling connect() or init().
+    *
+    * @note Derived classes must use this constructor and call init() or
+    * connect() themselves. Derived classes should also call deinit() in
+    * their destructor. This ensures that no interrupts can occur when the
+    * object is partially constructed or destroyed.
+    *
+    * @param phy USB phy to use
+    * @param frequency_rx frequency in Hz (default: 48000)
+    * @param channel_count_rx channel number (1 or 2) (default: 1)
+    * @param frequency_tx frequency in Hz (default: 8000)
+    * @param channel_count_tx channel number (1 or 2) (default: 1)
+    * @param buffer_ms time audio can be buffered without overflowing in milliseconds
+    * @param vendor_id Your vendor_id
+    * @param product_id Your product_id
+    * @param product_release Your product_release
+    */
+    USBAudio(USBPhy *phy, uint32_t frequency_rx, uint8_t channel_count_rx, uint32_t frequency_tx, uint8_t channel_count_tx, uint32_t buffer_ms, uint16_t vendor_id, uint16_t product_id, uint16_t product_release);
+
+    /**
+     * Destroy this object
+     *
+     * Any classes which inherit from this class must call deinit
+     * before this destructor runs.
+     */
+    virtual ~USBAudio();
+
+    /**
+    * Connect USBAudio
+    */
+    void connect();
+
+    /**
+    * Disconnect USBAudio
+    *
+    * This unblocks all calls to read_ready and write_ready.
+    */
+    void disconnect();
+
+    /**
+    * Read audio data
+    *
+    * @param buf pointer on a buffer which will be filled with audio data
+    * @param size size to read
+    *
+    * @returns true if successful
+    */
+    bool read(uint8_t *buf, uint32_t size);
+
+    /**
+    * Nonblocking audio data read
+    *
+    * Read the available audio data.
+    *
+    * @param buf pointer on a buffer which will be filled with audio data
+    * @param size size to read
+    * @param actual size actually read
+    * @note This function is safe to call from USBAudio callbacks.
+    *
+    * @returns true if successful
+    */
+    void read_nb(uint8_t *buf, uint32_t size, uint32_t *actual);
```

Comment:
I update the description as that was a mistake.

As for the return value, I was trying to keep this compatible with the `read` function. This leaves open the option to change the return value to `bool` to return false if the write failed because the device is out of the configured state (as `read` does). Do you think this is reasonable to keep the same?

Reply:
In that case, I think `read_nb()` can actually return `bool` value in addition to updating `actual`, so that the user of this API can decide to verify any of them.

Reply:
Thanks for pointing out an analogous function in USBCDC https://github.com/ARMmbed/mbed-os/blob/4fedfbb974e6211c0e20a775da9a6840a64c265c/usb/device/USBSerial/USBCDC.h#L126 Updating `actual` through a pointer looks more consistent now.